File: problems/Coin Change/coinChange.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        // return coinChangeBfs(coins, amount);
        return coinChangeBottomUp(coins, amount);
    }

    int coinChangeBottomUp(int[] coins, int amount) {
        int[] dp = new int[1 + amount];

        Arrays.fill(dp, 1 + amount);
        dp[0] = 0;

        for (int idx = 0; coins.length > idx; ++idx) {
            for (int sum = coins[idx]; amount >= sum; ++sum) {
                dp[sum] = Math.min(dp[sum], 1 + dp[sum - coins[idx]]);
            }
        }

        return 1 + amount == dp[amount] ? -1 : dp[amount];
    }

    int coinChangeBfs(int[] coins, int amount) {
        int result = 0;
        Queue<Integer> bfs = new LinkedList<>();
        HashSet<Integer> visited = new HashSet<>();

        if (0 >= amount) {
            return result;
        }

        bfs.add(amount);

        while (!bfs.isEmpty()) {
            ++result;

            for (int idx = bfs.size(); 0 < idx; --idx) {
                int item = bfs.poll();

                for (int coin : coins) {
                    if (item == coin) {
                        return result;
                    } else if (item > coin && !visited.contains(item - coin)) {
                        bfs.add(item - coin);
                        visited.add(item - coin);
                    }
                }
            }
        }

        return -1;
    }
}

File: problems/Coin Change/coinChange.cs
public class Solution {
    public int CoinChange(int[] coins, int amount) {
        return CoinChangeBottomUp(coins, amount);
    }

    public int CoinChangeBottomUp(int[] coins, int amount) {
        int[] dp = new int[1 + amount];

        Array.Fill(dp, int.MaxValue);

        foreach (int coin in coins) {
            for (int sum = 0; amount >= sum; ++sum) {
                if (0 == sum) {
                    dp[sum] = 0;
                } else if (coin <= sum) {
                    var sumWithout = dp[sum];
                    var sumWith = int.MaxValue == dp[sum - coin] ? int.MaxValue : 1 + dp[sum - coin];

                    dp[sum] = Math.Min(sumWith, sumWithout);
                }
            }
        }

        return int.MaxValue == dp[amount] ? -1 : dp[amount];
    }
}
